sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java

```diff
@@ -257,42 +306,56 @@ public void populateDisplayData(DisplayData.Builder builder) {
    */
   private class WriteShardedBundles<WriteT> extends DoFn<KV<Integer, Iterable<T>>, WriteT> {
     private final PCollectionView<WriteOperation<T, WriteT>> writeOperationView;
+    private final PCollectionView<Integer> numShardsView;
```

Comment:
The refactor to take numShards as a side input might also be an easy preparatory change. Not necessarily needed now, since this should go in soon. But given that some changes I think you've made don't seem to be here, I'm worried we've gotten branches mixed up somehow. It might be worth getting pieces of this in.